pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,363 @@
+	masterIgnAsset := &machine.Master{}
+	parents.Get(
+		capiManifestsAsset,
+		clusterKubeconfigAsset,
+		clusterID,
+		installConfig,
+		bootstrapIgnAsset,
+		masterIgnAsset,
+		capiMachinesAsset,
+	)
+
+	fileList := []*asset.File{}
+
+	// Collect cluster and non-machine-related infra manifests
+	// to be applied in the first stage.
+	infraManifests := []client.Object{}
+	for _, m := range capiManifestsAsset.RuntimeFiles() {
+		infraManifests = append(infraManifests, m.Object)
+	}
+
+	machineManifests := []client.Object{}
+	for _, m := range capiMachinesAsset.RuntimeFiles() {
+		machineManifests = append(machineManifests, m.Object)
+	}
+
+	// TODO(vincepri): The context should be passed down from the caller,
+	// although today the Asset interface doesn't allow it, refactor once it does.
+	ctx, cancel := context.WithCancel(signals.SetupSignalHandler())
+	defer cancel()
+	go func() {
+		<-ctx.Done()
+		cancel()
+		clusterapi.System().Teardown()
+	}()
+
+	preProvisionInput := PreProvisionInput{
+		InfraID:       clusterID.InfraID,
+		InstallConfig: installConfig,
+	}
+
+	if p, ok := i.capiProvider.(PreProvider); ok {
+		if err := p.PreProvision(ctx, preProvisionInput); err != nil {
+			return fileList, fmt.Errorf("failed during pre-provisioning: %w", err)
+		}
+	} else {
+		logrus.Debugf("No pre-provisioning requirements for the %s provider", i.capiProvider.Name())
+	}
+
+	// Run the CAPI system.
+	capiSystem := clusterapi.System()
+	if err := capiSystem.Run(ctx, installConfig); err != nil {
+		return fileList, fmt.Errorf("failed to run cluster api system: %w", err)
+	}
+
+	// Grab the client.
+	cl := capiSystem.Client()
+
+	// Create the infra manifests.
+	for _, m := range infraManifests {
+		m.SetNamespace(capiutils.Namespace)
+		if err := cl.Create(context.Background(), m); err != nil {
+			return fileList, fmt.Errorf("failed to create manifest: %w", err)
+		}
+		logrus.Infof("Created manifest %+T, namespace=%s name=%s", m, m.GetNamespace(), m.GetName())
+	}
+
+	// Pass cluster kubeconfig and store it in; this is usually the role of a bootstrap provider.
+	{
+		key := client.ObjectKey{
+			Name:      clusterID.InfraID,
+			Namespace: capiutils.Namespace,
+		}
+		cluster := &clusterv1.Cluster{}
+		if err := cl.Get(context.Background(), key, cluster); err != nil {
+			return fileList, err
+		}
+		// Create the secret.
+		clusterKubeconfig := clusterKubeconfigAsset.Files()[0].Data
+		secret := utilkubeconfig.GenerateSecret(cluster, clusterKubeconfig)
+		if err := cl.Create(context.Background(), secret); err != nil {
+			return fileList, err
+		}
+	}
+
+	// Wait for the load balancer to be ready by checking the control plane endpoint
+	// on the cluster object.
+	var cluster *clusterv1.Cluster
+	{
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Duration: time.Second * 10,
+			Factor:   float64(1.5),
+			Steps:    32,
+		}, func() (bool, error) {
+			c := &clusterv1.Cluster{}
+			if err := cl.Get(context.Background(), client.ObjectKey{
```

Comment:
this line could use the context defined line 147
```suggestion
			if err := cl.Get(ctx, client.ObjectKey{
```